game_patch/multi/multi.cpp

```diff
@@ -266,10 +280,14 @@ std::pair<bool, float> multi_is_cps_above_limit(rf::Player* pp, float max_cps)
 float multi_get_max_cps(int weapon_type, bool alt_fire)
 {
     if (rf::weapon_is_semi_automatic(weapon_type)) {
-        // most people cannot do much more than 10 clicks per second so 20 should be safe
-        return 20.0f;
+        // since semi-auto weapons can be fired as quickly as click input, limit max allowed rate to value set by server, default 20/sec
+        return g_additional_server_config.click_limiter_max_cps;
+    }
+    if (rf::weapon_is_riot_stick(weapon_type)) {
+        // double (+ 10%) baton max fire rate to fix legit fire requests being dropped when holding primary fire
```

Comment:
I guess you measured it experimentally but you don't know why this formula works?
It would be nice to understand how the game calculates it. I am worried that it won't work in case of some mods that customize riot stick

Reply:
I haven't traced it back, but from testing I am virtually 100% sure it's caused by the baton "firing twice" due to having 2 $Impact Delay entries in weapons.tbl. The baton is the only player weapon in the stock game like that.

Perhaps a better solution could be to read the number of $Impact Delay entries in the weapon's weapons.tbl entry and multiply the fire wait by that number before feeding it into the max_cps calculation? I'm just not sure off-hand how simple it will be to get that information.

> I am worried that it won't work in case of some mods that customize riot stick
I can't see a scenario where just doubling the allowable fire rate for the baton would break anything in the case of a mod, though it could in theory open some doors for people to cheat by rapid firing a baton in a mod (which while not ideal, is a _very_ low risk based on both likelihood and impact).

For reference, I show the issue with the baton in the video I linked in a comment below.

Reply:
I thought about it for a while and adjusted the approach to resolve this, now it simply excludes melee weapons from the click limiter.

This isn't an absolutely ideal solution, but I think it's a good compromise. While yes someone _could_ cheat by rapid firing a melee weapon, the impact of that type of cheat is so incredibly low that I'm not sure it's worth the risk or effort associated with developing a way of handling it "properly". It's also fairly common for mods to change baton functionality substantially, so if we didn't want to exclude melee weapons, we'd have to figure out a good way to handle all reasonable possibilities.

All in all, given the _extremely_ low impact of a potential rapid fire cheat for melee weapons, I'm thinking it's just safer to exclude them and not worry about breaking things. It also solves the baton issue in the current build.